compiler/plugins/target/ROCM/Dialect/ROCM/Transforms/ApplyBuiltinPDLPatterns.cpp

```diff
@@ -251,4 +263,86 @@ class ApplyBuiltinPDLPatternsPass
   FrozenRewritePatternSet patterns;
 };
 
+class ApplyBuiltinPDLPatternsDriverPass final
+    : public mlir::iree_compiler::IREE::ROCM::impl::
+          ApplyBuiltinPDLPatternsDriverPassBase<
+              ApplyBuiltinPDLPatternsDriverPass> {
+public:
+  using Base::Base;
+
+  void runOnOperation() override {
+    ModuleOp moduleOp = getOperation();
+    IREE::GPU::TargetAttr targetAttr = getGPUTargetAttr(moduleOp);
+    if (!targetAttr) {
+      moduleOp->emitOpError() << "no executable target attribute found.";
+      return signalPassFailure();
+    }
+
+    IREE::ROCM::ApplyBuiltinPDLPatternsPassOptions options;
+    options.enableTensorUKernels = true;
+    options.targets.push_back(targetAttr.getArch().str());
+    OpPassManager passManager(moduleOp.getOperationName());
+    FunctionLikeNest(passManager).addPass([&]() {
+      return createApplyBuiltinPDLPatternsPass(options);
+    });
+    if (failed(runPipeline(passManager, moduleOp))) {
+      signalPassFailure();
+    }
+
+    // Parse the required builtin ukernels and embed them into the module.
+    MLIRContext *ctx = moduleOp.getContext();
+    auto rocmDialect = ctx->getOrLoadDialect<IREE::ROCM::ROCMDialect>();
+    SmallVector<IREE::Util::FuncOp> ukernelFunctions;
+    auto res = moduleOp.walk([&](Operation *op) {
+      auto builtinName =
+          dyn_cast_or_null<StringAttr>(op->getAttr(kBuiltinName));
+      auto ukernelDesc = dyn_cast_or_null<IREE::Codegen::UKernelDescriptorAttr>(
+          op->getAttr(kUKernelDescriptorName));
+      if (!builtinName || !ukernelDesc) {
+        return WalkResult::advance();
+      }
+      if (moduleOp->hasAttr(ukernelDesc.getUkernelName())) {
+        // Avoid parsing and serializing the same ukernel again and again.
+        return WalkResult::advance();
+      }
+      std::optional<StringRef> maybeBuiltin =
+          rocmDialect->getBuiltin(builtinName.str());
+      if (!maybeBuiltin) {
+        moduleOp.emitOpError()
+            << "could not find builtin with name " << builtinName.str();
+        return WalkResult::interrupt();
+      }
+      OwningOpRef<ModuleOp> builtinModule =
+          parseSourceString<ModuleOp>(*maybeBuiltin, ctx);
+      auto *symbolTableOp =
+          SymbolTable::getNearestSymbolTable(builtinModule.get());
+      SymbolTable symbolTable(symbolTableOp);
+      auto funcOp = dyn_cast_if_present<IREE::Util::FuncOp>(
```

Comment:
Can this be `FunctionOpInterface` or just `Operation*`? Since you're parsing the full module I don't think this runs into the issue we discussed the other day.

Reply:
Yeah, it can be a FunctionOpInterface again now. Updated it.